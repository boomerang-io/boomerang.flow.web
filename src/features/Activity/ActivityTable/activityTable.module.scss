@import "~@boomerang/boomerang-styles/lib/colors";
@import "~Styles/variables";
@import "~Styles/mixins";

.tableSkeleton {
  th {
    color: #047cc0;
  }
  thead {
    background-color: #d5d9e0;
    th:nth-of-type(1) {
      width: 15% !important;
    }

    th:nth-of-type(2) {
      width: 25% !important;
    }

    th:nth-of-type(3) {
      width: 10% !important;
    }

    th:nth-of-type(4) {
      width: 15% !important;
    }

    th:nth-of-type(5) {
      width: 15% !important;
    }

    th:nth-of-type(6) {
      width: 10% !important;
    }

    th:nth-of-type(7) {
      width: 10% !important;
    }
  }
}

.tableContainer {
  margin-top: 1rem;
  margin-bottom: 2rem;
}

.header {
  display: flex;
  justify-content: space-between;
}

.tableHeadHeader {
  color: #171717;
  font-size: 0.875rem;
  font-weight: 600;
  height: 3rem;
  letter-spacing: 0.01rem;
  line-height: 1.125rem;

  button {
    width: 100% !important;
  }

  &.teamName {
    width: 15%;
  }

  &.workflowName {
    width: 25%;
  }

  &.trigger {
    width: 10%;
  }

  &.initiatedByUserName {
    width: 15%;
  }

  &.creationDate {
    width: 15%;
  }

  &.duration {
    width: 10%;
  }

  &.status {
    width: 10%;
  }
}

.tableBody {
  font-size: 0.875rem;
  letter-spacing: 0.01rem;
  line-height: 1.125rem;
}

.tableRow {
  cursor: pointer;

  &:focus {
    @include focus-outline(var(--cds-interactive-01), -0.125rem);
  }

  &.completed {
    .status {
      color: #009c98;
      svg {
        fill: #009c98;
      }
    }
  }

  &.failure {
    .tableCell {
      background-color: #fcd0d3;
    }
    .status {
      color: #a51920;
      svg {
        fill: #a51920;
      }
    }
  }

  &.invalid {
    .tableCell {
      background-color: #b9bfc7;
    }
  }

  &.invalid,
  &.skipped {
    color: #242a2e;
    .status {
      color: #242a2e;
      svg {
        fill: #242a2e;
      }
    }
  }

  &.notstarted {
    .status {
      color: #50565b;
      svg {
        fill: #50565b;
      }
    }
  }

  &.inProgress {
    .tableCell {
      animation: inProgressText 3.75s infinite;
      svg {
        animation: inProgressIcon 3.75s infinite;
      }
    }
  }
}

@keyframes inProgressText {
  0% {
    color: $cool-gray-70;
  }
  20% {
    color: $aqua-60;
  }
  40% {
    color: $teal-50;
  }
  60% {
    color: $blue-50;
  }
  80% {
    color: $cool-gray-40;
  }
  100% {
    color: $cool-gray-70;
  }
}

@keyframes inProgressIcon {
  0% {
    fill: $cool-gray-70;
  }
  20% {
    fill: $aqua-60;
  }
  40% {
    fill: $teal-50;
  }
  60% {
    fill: $blue-50;
  }
  80% {
    fill: $cool-gray-40;
  }
  100% {
    fill: $cool-gray-70;
  }
}

.tableCell {
  align-items: center;
  color: #565656;
  display: flex;
  height: 3.625rem;
  line-height: 1.2;
  overflow-y: auto;
  padding-left: 0.75rem;
  width: 100%;

  .status {
    align-items: center;
    display: flex;
    .statusIcon {
      height: 1rem;
      width: 1rem;
      margin-right: 0.5rem;
    }
  }
}

.tableTextarea {
  background: inherit;
  margin-left: 0.25rem;
  max-width: 100%;
  resize: none;
  text-align: left;
  white-space: normal;
  width: 90%;
  word-break: break-all;
}
