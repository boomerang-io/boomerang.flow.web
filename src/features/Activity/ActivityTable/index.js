import React from "react";
import PropTypes from "prop-types";
import moment from "moment";
import cx from "classnames";
import { settings } from "carbon-components";
import queryString from "query-string";
import getHumanizedDuration from "@boomerang/boomerang-utilities/lib/getHumanizedDuration";
import isAccessibleEvent from "@boomerang/boomerang-utilities/lib/isAccessibleEvent";
import { DataTableSkeleton, DataTable, Pagination } from "carbon-components-react";
import { Error404 } from "@boomerang/carbon-addons-boomerang-react";
import { ACTIVITY_STATUSES_TO_TEXT, ACTIVITY_STATUSES_TO_ICON } from "Constants/activityStatuses";
import styles from "./activityTable.module.scss";

const { prefix } = settings;

ActivityTable.propTypes = {
  isLoading: PropTypes.bool,
  history: PropTypes.object.isRequired,
  location: PropTypes.object.isRequired,
  match: PropTypes.object.isRequired,
  tableData: PropTypes.object,
  updateHistorySearch: PropTypes.func.isRequired
};

const PAGE_SIZES = [5, 10, 20, 25, 50, 100];

const headers = [
  {
    header: "Team",
    key: "teamName"
  },
  {
    header: "Workflow",
    key: "workflowName"
  },
  {
    header: "Trigger",
    key: "trigger"
  },
  {
    header: "Initiated By",
    key: "initiatedByUserName"
  },
  {
    header: "Start Time",
    key: "creationDate"
  },
  {
    header: "Duration",
    key: "duration"
  },
  {
    header: "Status",
    key: "status"
  }
];

function renderCell(cellIndex, value) {
  const column = headers[cellIndex];

  switch (column.header) {
    case "Trigger":
      return (
        <p className={styles.tableTextarea} style={{ textTransform: "capitalize" }}>
          {value || "---"}
        </p>
      );
    case "Start Time":
      return <time className={styles.tableTextarea}>{moment(value).format("YYYY-MM-DD hh:mm A")}</time>;
    case "Duration":
      return (
        <time className={styles.tableTextarea}>{value ? getHumanizedDuration(parseInt(value / 1000, 10)) : "---"}</time>
      );
    case "Status":
      const Icon = ACTIVITY_STATUSES_TO_ICON[value ? value : "notstarted"];
      return (
        <div className={`${styles.status} ${styles[value]}`}>
          <Icon aria-label={value} className={styles.statusIcon} />
          <p className={styles.statusText}>{ACTIVITY_STATUSES_TO_TEXT[value ? value : "notstarted"]}</p>
        </div>
      );
    default:
      return <p className={styles.tableTextarea}>{value || "---"}</p>;
  }
}

function ActivityTable(props) {
  function handlePaginationChange({ page, pageSize }) {
    props.updateHistorySearch({
      ...queryString.parse(props.location.search),
      page: page - 1, // We have to decrement by one to offset the table pagination adjustment
      size: pageSize
    });
  }

  function handleSort(e, { sortHeaderKey }) {
    const { property, direction } = props.tableData.pageable.sort[0];
    const sort = sortHeaderKey;
    let order = "ASC";

    if (sort === property && direction === "ASC") {
      order = "DESC";
    }

    props.updateHistorySearch({ ...queryString.parse(props.location.search), sort, order });
  }

  function executionViewRedirect(activityId) {
    const activity = props.tableData.records.find(activity => activity.id === activityId);
    props.history.push({
      pathname: `/activity/${activity.workflowId}/execution/${activity.id}`,
      state: { fromUrl: `${props.match.url}${props.location.search}`, fromText: "Activity" }
    });
  }

  if (props.isLoading) {
    return (
      <div style={{ marginTop: "1rem" }}>
        <DataTableSkeleton
          className={cx(`${prefix}--skeleton`, `${prefix}--data-table`, styles.tableSkeleton)}
          rowCount={10}
          columnCount={headers.length}
          headers={headers.map(header => header.header)}
        />
      </div>
    );
  }

  const {
    pageable: { number, size, sort, totalElements },
    records
  } = props.tableData;
  const { TableContainer, Table, TableHead, TableRow, TableBody, TableCell, TableHeader } = DataTable;

  return (
    <section>
      <div className={styles.tableContainer}>
        {totalElements > 0 ? (
          <>
            <DataTable
              rows={records}
              headers={headers}
              render={({ rows, headers, getHeaderProps }) => (
                <TableContainer>
                  <Table>
                    <TableHead>
                      <TableRow>
                        {headers.map(header => (
                          <TableHeader
                            id={header.key}
                            {...getHeaderProps({
                              header,
                              className: `${styles.tableHeadHeader} ${styles[header.key]}`,
                              isSortable: header.header !== "Team" && header.header !== "Workflow",
                              onClick: handleSort
                            })}
                            isSortHeader={sort[0].property === header.key}
                            sortDirection={sort[0].direction}
                          >
                            {header.header}
                          </TableHeader>
                        ))}
                      </TableRow>
                    </TableHead>
                    <TableBody className={styles.tableBody}>
                      {rows.map(row => (
                        <TableRow
                          key={row.id}
                          className={`${styles.tableRow} ${styles[row.cells[6].value]}`}
                          data-testid="configuration-property-table-row"
                          onClick={() => executionViewRedirect(row.id)}
                          onKeyDown={e => isAccessibleEvent(e) && executionViewRedirect(row.id)}
                          tabIndex={0}
                        >
                          {row.cells.map((cell, cellIndex) => (
                            <TableCell key={cell.id} style={{ padding: "0" }}>
                              <div className={styles.tableCell}>{renderCell(cellIndex, cell.value)}</div>
                            </TableCell>
                          ))}
                        </TableRow>
                      ))}
                    </TableBody>
                  </Table>
                </TableContainer>
              )}
            />
            <Pagination
              onChange={handlePaginationChange}
              page={number + 1} // We need to offset by one bc service is zero indexed
              pageSize={size}
              pageSizes={PAGE_SIZES}
              totalItems={totalElements}
            />
          </>
        ) : (
          <>
            <DataTable
              rows={[]}
              headers={headers}
              render={({ headers }) => (
                <TableContainer>
                  <Table>
                    <TableHead>
                      <TableRow className={styles.tableHeadRow}>
                        {headers.map(header => (
                          <TableHeader
                            key={header.key}
                            id={header.key}
                            className={`${styles.tableHeadHeader} ${styles[header.key]}`}
                          >
                            {header.header}
                          </TableHeader>
                        ))}
                      </TableRow>
                    </TableHead>
                  </Table>
                </TableContainer>
              )}
            />
            <Error404
              message={"Execute some Workflows or try changing the filters"}
              title="No activity found"
              header={null}
            />
          </>
        )}
      </div>
    </section>
  );
}

export default ActivityTable;
