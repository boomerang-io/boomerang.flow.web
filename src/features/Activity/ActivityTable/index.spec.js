import React from "react";
import ActivityTable from "./index";

const props = {
  activities: [],
  isUpdating: false,
  sort: {},
  tableData: { records: [], pageable: { number: 0, size: 10, sort: "asc", totalElements: 10 } }
};

describe("ActivityTable --- Snapshot", () => {
  it("Capturing Snapshot of ActivityTable", () => {
    const { baseElement } = rtlRouterRender(<ActivityTable {...props} />);
    expect(baseElement).toMatchSnapshot();
  });
});
