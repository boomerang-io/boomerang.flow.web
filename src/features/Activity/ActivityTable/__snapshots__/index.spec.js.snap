// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActivityTable --- Snapshot Capturing Snapshot of ActivityTable 1`] = `
<body
  id="app"
>
  <div>
    <section>
      <div
        class="tableContainer"
      >
        <div
          class="bx--data-table-container"
        >
          <table
            class="bx--data-table bx--data-table--no-border"
          >
            <thead>
              <tr>
                <th
                  class="tableHeadHeader teamName"
                  id="teamName"
                  scope="col"
                >
                  <span
                    class="bx--table-header-label"
                  >
                    Team
                  </span>
                </th>
                <th
                  class="tableHeadHeader workflowName"
                  id="workflowName"
                  scope="col"
                >
                  <span
                    class="bx--table-header-label"
                  >
                    Workflow
                  </span>
                </th>
                <th
                  aria-sort="none"
                  class="tableHeadHeader trigger"
                  scope="col"
                >
                  <button
                    class="tableHeadHeader trigger bx--table-sort"
                    id="trigger"
                  >
                    <span
                      class="bx--table-header-label"
                    >
                      Trigger
                    </span>
                    <svg
                      aria-label="Sort rows by this header in ascending order"
                      class="bx--table-sort__icon"
                      focusable="false"
                      height="20"
                      preserveAspectRatio="xMidYMid meet"
                      role="img"
                      style="will-change: transform;"
                      viewBox="0 0 32 32"
                      width="20"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M16 4L6 14l1.41 1.41L15 7.83V30h2V7.83l7.59 7.58L26 14 16 4z"
                      />
                    </svg>
                    <svg
                      aria-label="Sort rows by this header in ascending order"
                      class="bx--table-sort__icon-unsorted"
                      focusable="false"
                      height="20"
                      preserveAspectRatio="xMidYMid meet"
                      role="img"
                      style="will-change: transform;"
                      viewBox="0 0 32 32"
                      width="20"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M26.59 20.59L23 24.17V4h-2v20.17l-3.59-3.58L16 22l6 6 6-6-1.41-1.41zM10 4l-6 6 1.41 1.41L9 7.83V28h2V7.83l3.59 3.58L16 10l-6-6z"
                      />
                    </svg>
                  </button>
                </th>
                <th
                  aria-sort="none"
                  class="tableHeadHeader initiatedByUserName"
                  scope="col"
                >
                  <button
                    class="tableHeadHeader initiatedByUserName bx--table-sort"
                    id="initiatedByUserName"
                  >
                    <span
                      class="bx--table-header-label"
                    >
                      Initiated By
                    </span>
                    <svg
                      aria-label="Sort rows by this header in ascending order"
                      class="bx--table-sort__icon"
                      focusable="false"
                      height="20"
                      preserveAspectRatio="xMidYMid meet"
                      role="img"
                      style="will-change: transform;"
                      viewBox="0 0 32 32"
                      width="20"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M16 4L6 14l1.41 1.41L15 7.83V30h2V7.83l7.59 7.58L26 14 16 4z"
                      />
                    </svg>
                    <svg
                      aria-label="Sort rows by this header in ascending order"
                      class="bx--table-sort__icon-unsorted"
                      focusable="false"
                      height="20"
                      preserveAspectRatio="xMidYMid meet"
                      role="img"
                      style="will-change: transform;"
                      viewBox="0 0 32 32"
                      width="20"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M26.59 20.59L23 24.17V4h-2v20.17l-3.59-3.58L16 22l6 6 6-6-1.41-1.41zM10 4l-6 6 1.41 1.41L9 7.83V28h2V7.83l3.59 3.58L16 10l-6-6z"
                      />
                    </svg>
                  </button>
                </th>
                <th
                  aria-sort="none"
                  class="tableHeadHeader creationDate"
                  scope="col"
                >
                  <button
                    class="tableHeadHeader creationDate bx--table-sort"
                    id="creationDate"
                  >
                    <span
                      class="bx--table-header-label"
                    >
                      Start Time
                    </span>
                    <svg
                      aria-label="Sort rows by this header in ascending order"
                      class="bx--table-sort__icon"
                      focusable="false"
                      height="20"
                      preserveAspectRatio="xMidYMid meet"
                      role="img"
                      style="will-change: transform;"
                      viewBox="0 0 32 32"
                      width="20"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M16 4L6 14l1.41 1.41L15 7.83V30h2V7.83l7.59 7.58L26 14 16 4z"
                      />
                    </svg>
                    <svg
                      aria-label="Sort rows by this header in ascending order"
                      class="bx--table-sort__icon-unsorted"
                      focusable="false"
                      height="20"
                      preserveAspectRatio="xMidYMid meet"
                      role="img"
                      style="will-change: transform;"
                      viewBox="0 0 32 32"
                      width="20"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M26.59 20.59L23 24.17V4h-2v20.17l-3.59-3.58L16 22l6 6 6-6-1.41-1.41zM10 4l-6 6 1.41 1.41L9 7.83V28h2V7.83l3.59 3.58L16 10l-6-6z"
                      />
                    </svg>
                  </button>
                </th>
                <th
                  aria-sort="none"
                  class="tableHeadHeader duration"
                  scope="col"
                >
                  <button
                    class="tableHeadHeader duration bx--table-sort"
                    id="duration"
                  >
                    <span
                      class="bx--table-header-label"
                    >
                      Duration
                    </span>
                    <svg
                      aria-label="Sort rows by this header in ascending order"
                      class="bx--table-sort__icon"
                      focusable="false"
                      height="20"
                      preserveAspectRatio="xMidYMid meet"
                      role="img"
                      style="will-change: transform;"
                      viewBox="0 0 32 32"
                      width="20"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M16 4L6 14l1.41 1.41L15 7.83V30h2V7.83l7.59 7.58L26 14 16 4z"
                      />
                    </svg>
                    <svg
                      aria-label="Sort rows by this header in ascending order"
                      class="bx--table-sort__icon-unsorted"
                      focusable="false"
                      height="20"
                      preserveAspectRatio="xMidYMid meet"
                      role="img"
                      style="will-change: transform;"
                      viewBox="0 0 32 32"
                      width="20"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M26.59 20.59L23 24.17V4h-2v20.17l-3.59-3.58L16 22l6 6 6-6-1.41-1.41zM10 4l-6 6 1.41 1.41L9 7.83V28h2V7.83l3.59 3.58L16 10l-6-6z"
                      />
                    </svg>
                  </button>
                </th>
                <th
                  aria-sort="none"
                  class="tableHeadHeader status"
                  scope="col"
                >
                  <button
                    class="tableHeadHeader status bx--table-sort"
                    id="status"
                  >
                    <span
                      class="bx--table-header-label"
                    >
                      Status
                    </span>
                    <svg
                      aria-label="Sort rows by this header in ascending order"
                      class="bx--table-sort__icon"
                      focusable="false"
                      height="20"
                      preserveAspectRatio="xMidYMid meet"
                      role="img"
                      style="will-change: transform;"
                      viewBox="0 0 32 32"
                      width="20"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M16 4L6 14l1.41 1.41L15 7.83V30h2V7.83l7.59 7.58L26 14 16 4z"
                      />
                    </svg>
                    <svg
                      aria-label="Sort rows by this header in ascending order"
                      class="bx--table-sort__icon-unsorted"
                      focusable="false"
                      height="20"
                      preserveAspectRatio="xMidYMid meet"
                      role="img"
                      style="will-change: transform;"
                      viewBox="0 0 32 32"
                      width="20"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M26.59 20.59L23 24.17V4h-2v20.17l-3.59-3.58L16 22l6 6 6-6-1.41-1.41zM10 4l-6 6 1.41 1.41L9 7.83V28h2V7.83l3.59 3.58L16 10l-6-6z"
                      />
                    </svg>
                  </button>
                </th>
              </tr>
            </thead>
            <tbody
              aria-live="polite"
              class="tableBody"
            />
          </table>
        </div>
        <div
          class="bx--pagination"
        >
          <div
            class="bx--pagination__left"
          >
            <label
              class="bx--pagination__text"
              for="bx-pagination-select-1"
              id="bx-pagination-select-1-count-label"
            >
              Items per page:
            </label>
            <div
              class="bx--form-item"
            >
              <div
                class="bx--select bx--select--inline bx--select__item-count"
              >
                <div
                  class="bx--select-input--inline__wrapper"
                >
                  <div
                    class="bx--select-input__wrapper"
                  >
                    <select
                      class="bx--select-input"
                      id="bx-pagination-select-1"
                    >
                      <option
                        class="bx--select-option"
                        value="5"
                      >
                        5
                      </option>
                      <option
                        class="bx--select-option"
                        value="10"
                      >
                        10
                      </option>
                      <option
                        class="bx--select-option"
                        value="20"
                      >
                        20
                      </option>
                      <option
                        class="bx--select-option"
                        value="25"
                      >
                        25
                      </option>
                      <option
                        class="bx--select-option"
                        value="50"
                      >
                        50
                      </option>
                      <option
                        class="bx--select-option"
                        value="100"
                      >
                        100
                      </option>
                    </select>
                    <svg
                      aria-label="open list of options"
                      class="bx--select__arrow"
                      focusable="false"
                      height="16"
                      preserveAspectRatio="xMidYMid meet"
                      role="img"
                      style="will-change: transform;"
                      viewBox="0 0 16 16"
                      width="16"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M8 11L3 6l.7-.7L8 9.6l4.3-4.3.7.7z"
                      />
                      <title>
                        open list of options
                      </title>
                    </svg>
                  </div>
                </div>
              </div>
            </div>
            <span
              class="bx--pagination__text"
            >
              1–10 of 10 items
            </span>
          </div>
          <div
            class="bx--pagination__right"
          >
            <div
              class="bx--form-item"
            >
              <div
                class="bx--select bx--select--inline bx--select__page-number"
              >
                <label
                  class="bx--label bx--visually-hidden"
                  for="bx-pagination-select-3"
                >
                  Page number, of 1 pages
                </label>
                <div
                  class="bx--select-input--inline__wrapper"
                >
                  <div
                    class="bx--select-input__wrapper"
                  >
                    <select
                      class="bx--select-input"
                      id="bx-pagination-select-3"
                    >
                      <option
                        class="bx--select-option"
                        value="1"
                      >
                        1
                      </option>
                    </select>
                    <svg
                      aria-label="open list of options"
                      class="bx--select__arrow"
                      focusable="false"
                      height="16"
                      preserveAspectRatio="xMidYMid meet"
                      role="img"
                      style="will-change: transform;"
                      viewBox="0 0 16 16"
                      width="16"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M8 11L3 6l.7-.7L8 9.6l4.3-4.3.7.7z"
                      />
                      <title>
                        open list of options
                      </title>
                    </svg>
                  </div>
                </div>
              </div>
            </div>
            <span
              class="bx--pagination__text"
            >
              of 1 pages
            </span>
            <button
              aria-label="Previous page"
              class="bx--pagination__button bx--pagination__button--backward bx--pagination__button--no-index"
              disabled=""
              type="button"
            >
              <svg
                aria-hidden="true"
                focusable="false"
                height="24"
                preserveAspectRatio="xMidYMid meet"
                style="will-change: transform;"
                viewBox="0 0 32 32"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M19 23l-8-7 8-7v14z"
                />
              </svg>
            </button>
            <button
              aria-label="Next page"
              class="bx--pagination__button bx--pagination__button--forward bx--pagination__button--no-index"
              disabled=""
              type="button"
            >
              <svg
                aria-hidden="true"
                focusable="false"
                height="24"
                preserveAspectRatio="xMidYMid meet"
                style="will-change: transform;"
                viewBox="0 0 32 32"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M13 9l8 7-8 7V9z"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
    </section>
  </div>
</body>
`;
